python/tank/log.py

```diff
@@ -24,35 +25,100 @@ class LogManager(object):
     This class contains a series of methods to help standardize log output
     and access. Since Toolkit uses the standard python logging interface,
     you can manually configure and associate your logging if you like.
+
+    .. note:: This is a singleton class, so every time you instantiate it,
+              the same object is returned.
     """
-    # a global and standard rotating log file handler
-    # for writing generic toolkit logs to disk
-    _std_file_handler = None
 
-    @classmethod
-    def initialize_std_handler(cls, handler=logging.StreamHandler()):
+    # keeps track of the single instance of the class
+    __instance = None
+
+    def __new__(cls, *args, **kwargs):
+        """
+        Ensures only one instance of the log manager exists.
+        """
+        # create the instance if it hasn't been created already
+        if not cls.__instance:
+            # remember the instance so that no more are created
+            instance = super(LogManager, cls).__new__(
+                cls,
+                *args,
+                **kwargs
+            )
+
+            # a global and standard rotating log file handler
+            # for writing generic toolkit logs to disk
+            instance._std_file_handler = None
+            instance._std_file_handler_log_name = None
+
+            # the root logger, created at code init
+            instance._root_logger = logging.getLogger(constants.ROOT_LOGGER_NAME)
+
+            # check the TK_DEBUG flag at startup
+            if constants.DEBUG_LOGGING_ENV_VAR in os.environ:
+                instance.global_debug = True
+
+            cls.__instance = instance
+
+
+        return cls.__instance
+
+    def _set_global_debug(self, state):
         """
-        Convenience method that initializes a standard logger
+        Sets the state of the global debug in toolkit.
+        """
+        if state:
+            self.root_logger.setLevel(logging.DEBUG)
+            log.debug(
+                "Debug logging was enabled. To permanently enable it, "
+                "set the %s environment variable." % constants.DEBUG_LOGGING_ENV_VAR
+            )
+        else:
+            log.debug("Disabling debug logging.")
+            self.root_logger.setLevel(logging.INFO)
+
+    def _get_global_debug(self):
+        """
+        Returns whether the toolkit main logger has got
+        debug log turned on or not. Controlling this property
+        has a global impact and affects all services attached
+        to the logger.
+
+        .. note:: Debug logging is off by default.
+                  If you want to permanently enable debug logging,
+                  set the environment variable ``TK_DEBUG``.
+        """
+        return self.root_logger.isEnabledFor(logging.DEBUG)
+
+    global_debug = property(_get_global_debug, _set_global_debug)
+
+    def initialize_custom_handler(self, handler=logging.StreamHandler()):
```

Comment:
Awesome, I needed to do something similar in the desktop so that by adding --debug on the command line would log to a stream handler too.